examples/ransac_compare.py

```diff
@@ -0,0 +1,213 @@
+
+start_time = perf_counter()
+bad_by_ransac_pyprep, corr_pyprep = ransac_pyprep.find_bad_by_ransac(
+    raw._data.copy(),
+    raw.info["sfreq"],
+    len(raw.times),
+    np.asarray(raw.info["ch_names"]),
+    raw._get_channel_positions(),
+    [],
+    n_samples=n_samples,
+    fraction_good=fraction_good,
+    corr_thresh=corr_thresh,
+    fraction_bad=fraction_bad,
+    corr_window_secs=corr_window_secs,
+    channel_wise=False,
+    random_state=random_state,
+)
+print("--- %s seconds ---" % (perf_counter() - start_time))
+
+# Check channels that go bad together by correlation (RANSAC)
+print("pyprep bad chs:", bad_by_ransac_pyprep)
+assert set(bad_ch_names) == set(bad_by_ransac_pyprep)
+
+###############################################################################
+# Now we test the algorithms in a real EEG
+# Let's download some data for testing. Picking the 1st run of subject 4 here.
+data_paths = mne.datasets.eegbci.load_data(subject=4, runs=1, update_path=True)
+fname_test_file = data_paths[0]
+
+###############################################################################
+# Load data and prepare it
+# ------------------------
+
+raw = mne.io.read_raw_edf(fname_test_file, preload=True)
+
+# The eegbci data has non-standard channel names. We need to rename them:
+mne.datasets.eegbci.standardize(raw)
+
+# Add a montage to the data
+montage_kind = "standard_1005"
+montage = mne.channels.make_standard_montage(montage_kind)
+raw.set_montage(montage)
+
+
+###############################################################################
+# autoreject's RANSAC
+ransac_ar = Ransac(
+    picks=None,
+    n_resample=n_samples,
+    min_channels=fraction_good,
+    min_corr=corr_thresh,
+    unbroken_time=fraction_bad,
+    n_jobs=1,
+    random_state=random_state,
+    verbose="tqdm",
+)
+epochs = mne.make_fixed_length_epochs(
+    raw,
+    duration=corr_window_secs,
+    preload=True,
+    reject_by_annotation=False,
+    verbose=None,
+)
+
+start_time = perf_counter()
+ransac_ar = ransac_ar.fit(epochs)
+print("--- %s seconds ---" % (perf_counter() - start_time))
+
+corr_ar = ransac_ar.corr_
+bad_by_ransac_ar = ransac_ar.bad_chs_
+
+# Check channels that go bad together by correlation (RANSAC)
+print("autoreject bad chs:", bad_by_ransac_ar)
+
+
+###############################################################################
+# pyprep's RANSAC
+
+start_time = perf_counter()
+bad_by_ransac_pyprep, corr_pyprep = ransac_pyprep.find_bad_by_ransac(
+    raw._data.copy(),
+    raw.info["sfreq"],
+    len(raw.times),
+    np.asarray(raw.info["ch_names"]),
+    raw._get_channel_positions(),
+    [],
+    n_samples=n_samples,
+    fraction_good=fraction_good,
+    corr_thresh=corr_thresh,
+    fraction_bad=fraction_bad,
+    corr_window_secs=corr_window_secs,
+    channel_wise=False,
+    random_state=random_state,
+)
+print("--- %s seconds ---" % (perf_counter() - start_time))
+
+# Check channels that go bad together by correlation (RANSAC)
```

Comment:
re: comment --> "bad by correlation" was a different step to find bad channels, right? RANSAC is a bit more complex than going bad together by correlation, I thought.

Reply:
Yeah you are right, is a different step and is more complex. I guess the sole idea is them "going bad together" and using a correlation metric within a RANSAC framework; that is, is not just a simple correlation.


Will leave the original explanation of the paper here for reference so that I can redact a better explanation:

> The predictability criterion also relies on the channel
correlations of the low frequency portion of EEG signals.
Although the bad-by-correlation measure effectively detects most
bad channels, there are some situations in which channels “go
bad together.” We use the RANSAC (random sample consensus)
method (Fischler and Bolles, 1981) to select a random subset of
(so far) good channels to predict the behavior of each (excluded
from subset) channel in small non-overlapping time windows
(5s by default). The implementation is based on functions in
BCILAB (Kothe and Makeig, 2013).

> The bad-by-RANSAC channels behave in a manner poorly
predicted by the other channels. Before applying RANSAC,
we remove channels designated as noisy by other methods.
We then select a random subset of predictor channels for
each channel (25% by default). If not enough channels remain
to form the required subsets, the algorithm terminates. The
RANSAC algorithm uses a method of spherical splines for
estimating scalp potential based on algorithms proposed by
Perrin et al. (1989). Bad-by-RANSAC channels have a correlation
less than a threshold (0.75 by default) with their RANSAC-
predicted time courses on more than a certain fraction of
the windows (0.4 by default). The RANSAC default window
size is 4s.